networking/src/Network/Broadcast/OutboundQueue.hs

```diff
@@ -959,50 +977,81 @@ intDequeue outQ@OutQ{..} threadRegistry@TR{} sendMsg = do
     -- device), with no real way to prioritize any one thread over the other. We
     -- will be able to solve this conumdrum properly once we move away from TCP
     -- and use the RINA network architecture instead.
-    sendPacket :: EnqPacket msg nid -> m ()
-    sendPacket (EnqPacket p) = do
-      sendStartTime <- liftIO $ getCurrentTime
-
-      -- Mark the message as in-flight, limiting both enqueues and dequeues
-      setInFlightFor p (\n -> n + 1) qInFlight
-
-      -- We mark the node as rate limited, making it unavailable for any
-      -- additional dequeues until the timer expires and we mark it as
-      -- available again. We start the timer /before/ the send so that the
-      -- duration of the send does not affect when the next dequeue can take
-      -- places (apart from max-in-flight, of course).
-      case deqRateLimit $ qDequeuePolicy (packetDestType p) of
-        NoRateLimiting -> return ()
-        MaxMsgPerSec n -> do
-          let delay = 1000000 `div` n
-          applyMVar_ qRateLimited $ Set.insert (packetDestId p)
-          forkThread threadRegistry $ \unmask -> unmask $
-            (liftIO $ threadDelay delay) `finally` (liftIO $ do
-              applyMVar_ qRateLimited $ Set.delete (packetDestId p)
-              poke qSignal)
-
-      forkThread threadRegistry $ \unmask -> do
-        logDebug $ debugSending p
-
-        ma <- try $ unmask $ sendMsg (packetPayload p) (packetDestId p)
-
-        -- Reduce the in-flight count ..
-        setInFlightFor p (\n -> n - 1) qInFlight
-        liftIO $ poke qSignal
-
-        -- .. /before/ notifying the sender that the send is complete.
-        -- If we did this the other way around a subsequent enqueue might fail
-        -- because of policy restrictions on the max in-flight.
-        liftIO $ putMVar (packetSent p) ma
-
-        case ma of
-          Left err -> do
-            logFailure outQ FailedSend (Some p, err)
-            intFailure outQ p sendStartTime err
-          Right _  ->
-            return ()
-
-        logDebug $ debugSent p
+    sendPacket :: EnqPacket msg nid -> IO ()
+    sendPacket (EnqPacket p) = mdo
+      -- 😱  Does this even work?
+      -- We can't spawn the thread inside the STM transaction of course. So
+      -- instead we lazily fill it will the new status, using the old status
+      -- to determine what the new status shall be. In case the old status
+      -- is PacketEnqueued, computing the new status requires forking a
+      -- thread.
+      -- Should be fine I think.
+      oldStatus <- atomically $ do
+        status <- readTVar (packetStatus p)
+        writeTVar (packetStatus p) newStatus
+        return status
+      newStatus <- case oldStatus of
```

Comment:
This is a bit dodgy. I wasn't even sure it would work. The STM transaction is lazy in `newStatus`, so that's OK. But what happens if another thread reads that `packetStatus p` `TVar` and tries to force the value, in-between this transaction and this thread's determination of `newStatus`? I assume it'll just block until this thread has finished, but honestly I'm not sure. Seems to work though!

Reply:
That's interesting edge case. It seems it should be guaranteed by the recursive do rather than atomicity of STM.